### ISSUE TITLE
[Training/RL] Native HF compatibility contract, verified matrix, and DAPO/verl path

### ISSUE THREAD

**Issue by 123123213weqw** [rwkv-rs/hf-adapter#142]
## 目的

这条 issue 把当前仓库的**训练/RL 接口、已验证证据和新算法接入边界**写清楚，回答一个核心问题：Native 路径虽然有优化 kernel，但训练并不是只能走封闭的 fused 调用；它同时实现了标准 Hugging Face CausalLM contract，因此上层 Trainer/RL 框架可以直接驱动模型。

这里严格区分三种状态：

- **已验证（PASS）**：仓库已有可运行脚本、参数更新门禁和硬件记录。
- **接口兼容（contract-ready）**：模型侧所需接口已经具备，但该算法/框架还没有专门 smoke，不能写成“已验证”。
- **待适配**：框架内部 hard-code 了模型结构或 rollout engine，需要新增 adapter/backend。

## 1. Native 如何兼容训练框架

Native 模型没有把 RWKV 改成 Transformer Attention；它兼容的是 Transformers 的 API contract。

### 标准 HF 模型表面

[`NativeRWKV7ForCausalLM`](https://github.com/rwkv-rs/hf-adapter/blob/a37956d7ef616cbc7ecbcda0e95483ecf491f5ff/rwkv7_hf/native_model.py#L371-L417) 继承 `PreTrainedModel` 和 `GenerationMixin`，并提供：

- `from_pretrained()` / `save_pretrained()`；
- `get_input_embeddings()` / `get_output_embeddings()`；
- 标准 `forward(input_ids, attention_mask, labels, past_key_values, use_cache, ...)`；
- 标准 `CausalLMOutputWithPast(loss, logits, past_key_values, hidden_states)`；
- `generate()` 和 RWKV recurrent cache。

### SFT/普通 Trainer 路径

有 `labels` 时，Native 会：

1. 初始化每个 batch 的 RWKV state；
2. 选择可微 Native 路径（`use_jit=False`）；
3. `collect_all=True`，得到 `[B,T,V]` 的全部 token logits；
4. 执行标准 causal shift；
5. 使用 `F.cross_entropy(..., ignore_index=-100)`；
6. 返回 `CausalLMOutputWithPast.loss`。

实现见 [`native_model.py#L704-L763`](https://github.com/rwkv-rs/hf-adapter/blob/a37956d7ef616cbc7ecbcda0e95483ecf491f5ff/rwkv7_hf/native_model.py#L704-L763)。因此 HF Trainer/SFTTrainer 不需要认识 WKV 数学，只需要读取 `outputs.loss`。

### DPO/GRPO/在线 RL 路径

RL trainer 通常不把 policy loss 当作 `labels` CE 交给模型，而是要求模型返回每个 token 的 logits/log-prob，再由 trainer 计算 ratio、advantage、KL 和 policy loss。

Native 在无 cache 的完整前向中明确启用 `collect_all=True`，返回 `[B,T,V]`，代码注释也标明这是给 DPO/eval per-token log-prob 使用：[`native_model.py#L823-L853`](https://github.com/rwkv-rs/hf-adapter/blob/a37956d7ef616cbc7ecbcda0e95483ecf491f5ff/rwkv7_hf/native_model.py#L823-L853)。

因此它具备在线 RL 的两个模型侧基本接口：

```text
rollout: model.generate(..., use_cache=True)
actor update: model(..., use_cache=False).logits -> per-token log-prob -> RL loss
```

算法逻辑仍属于 TRL/verl/OpenRLHF 等 trainer，而不是写死在 RWKV kernel 中。

### PEFT/LoRA

RWKV projection/FFN 仍是标准 `nn.Linear` 模块，PEFT 可以按名称注入 LoRA，例如：

```text
r_proj, k_proj, v_proj, o_proj, key, value
```

所以无需为 RWKV 维护一套私有 LoRA 格式。

## 2. 当前已验证矩阵

权威状态页：[`docs/TRAINING.md`](https://github.com/rwkv-rs/hf-adapter/blob/main/docs/TRAINING.md)；可复制命令：[`docs/TRAINING_WORKFLOWS.md`](https://github.com/rwkv-rs/hf-adapter/blob/main/docs/TRAINING_WORKFLOWS.md)。

| 训练/框架 | 当前状态 | 仓库证据 |
|---|---|---|
| PyTorch causal LM backward | **PASS** | [`test_native_model_training_unit.py`](https://github.com/rwkv-rs/hf-adapter/blob/main/tests/test_native_model_training_unit.py) |
| HF Trainer | **PASS** | [`test_native_trainer_smoke.py`](https://github.com/rwkv-rs/hf-adapter/blob/main/tests/test_native_trainer_smoke.py) |
| Trainer checkpoint resume | **PASS** | [`test_native_trainer_resume_smoke.py`](https://github.com/rwkv-rs/hf-adapter/blob/main/tests/test_native_trainer_resume_smoke.py) |
| PEFT LoRA forward/backward | **PASS** | [`test_peft_lora.py`](https://github.com/rwkv-rs/hf-adapter/blob/main/tests/test_peft_lora.py) |
| PEFT save/load/merge | **PASS** | [`test_native_peft_save_load_merge.py`](https://github.com/rwkv-rs/hf-adapter/blob/main/tests/test_native_peft_save_load_merge.py) |
| TRL SFTTrainer | **PASS** | [`test_native_sft_smoke.py`](https://github.com/rwkv-rs/hf-adapter/blob/main/tests/test_native_sft_smoke.py) |
| TRL DPOTrainer | **PASS** | [`test_native_dpo_smoke.py`](https://github.com/rwkv-rs/hf-adapter/blob/main/tests/test_native_dpo_smoke.py) |
| TRL GRPOTrainer | **PASS** | [`test_native_grpo_smoke.py`](https://github.com/rwkv-rs/hf-adapter/blob/main/tests/test_native_grpo_smoke.py) |
| DeepSpeed ZeRO-2/3 | **PASS for documented smoke matrix** | [`test_deepspeed_training_smoke.py`](https://github.com/rwkv-rs/hf-adapter/blob/main/tests/test_deepspeed_training_smoke.py), [`test_deepspeed_resume_smoke.py`](https://github.com/rwkv-rs/hf-adapter/blob/main/tests/test_deepspeed_resume_smoke.py) |
| Official RWKV `train_temp` math | **PASS for documented exact lanes** | [`docs/TRAIN_TEMP_CUDA.md`](https://github.com/rwkv-rs/hf-adapter/blob/main/docs/TRAIN_TEMP_CUDA.md) |

硬件证据覆盖 V100、T4、A100、A800、RTX A6000、RTX 5090、Apple MPS 和 AMD ROCm 的不同子矩阵；这代表已声明形状上的兼容/更新证据，不应扩大解释为所有规模和所有分布式拓扑均已完成生产收敛：

- [V100](https://github.com/rwkv-rs/hf-adapter/blob/main/docs/validation/V100_HF_VALIDATION.md)
- [T4](https://github.com/rwkv-rs/hf-adapter/blob/main/docs/hardware/TURING_T4.md)
- [A100](https://github.com/rwkv-rs/hf-adapter/blob/main/docs/validation/A100_HF_VALIDATION.md)
- [A800](https://github.com/rwkv-rs/hf-adapter/blob/main/docs/validation/A800_HF_VALIDATION.md)
- [Apple Silicon](https://github.com/rwkv-rs/hf-adapter/blob/main/docs/hardware/APPLE_SILICON.md)
- [AMD ROCm](https://github.com/rwkv-rs/hf-adapter/blob/main/docs/validation/AMD_ROCM_HF_VALIDATION.md)

## 3. 其他训练/RL 算法如何接入

不是每出现一个新 RL 名字就改一次 RWKV 模型。应按算法需要的 model contract 分类。

### A. 直接监督/偏好优化

SFT、DPO、ORPO、CPO、KTO 等通常需要：

```text
标准 CausalLM forward + 完整 logits + attention/label mask + autograd
```

Native 已满足这个基础 contract。当前 **SFT/DPO 已验证**；ORPO/CPO/KTO 属于 **contract-ready，尚未逐个 smoke**。

### B. 在线 policy optimization

GRPO、RLOO、PPO、Online DPO、XPO 等通常还需要：

```text
批量 rollout + sampled token ids + per-token log-prob + reward + old/ref policy
```

Native 的 `generate()`、recurrent cache 和完整 differentiable logits 已满足模型侧基础 contract。当前 **GRPO 已验证**；其余算法需要增加各自 smoke 才能升级为 PASS。PPO 若使用 value head，还需验证/增加对应 value-head wrapper。

### C. DAPO

DAPO 不是新的模型结构，而是在 GRPO 类 pipeline 上加入：

1. asymmetric Clip-Higher；
2. dynamic group sampling/filtering；
3. token-level policy-gradient aggregation；
4. overlong reward shaping。

参考：[DAPO paper](https://arxiv.org/abs/2503.14476) 和 [official verl DAPO recipe](https://github.com/verl-project/verl-recipe/tree/main/dapo)。

建议提供两层支持：

- **TRL user path**：以现有 Native GRPO smoke 为起点，使用 `GRPOConfig(loss_type="dapo", epsilon=0.2, epsilon_high=0.28)`；补 overlong reward 与 dynamic group refill。该路径追求单机易用。
- **verl reproduction path**：模型以 HF repository + `trust_remote_code=True` 加载，首个正确性版本使用 FSDP actor 和 HF rollout。先设置 `use_remove_padding=False`，避免把不同样本拼接后污染 RWKV recurrent state。大规模性能版再增加 RWKV rollout worker。

在四项算法机制和参数更新都通过前，不能只因为设置了 `loss_type="dapo"` 就声称完整官方 DAPO 已复现。

## 4. 框架边界

| 框架层 | 状态 | 说明 |
|---|---|---|
| Transformers / Trainer | **已验证** | 标准 CausalLM loss/output contract |
| PEFT | **已验证** | 标准 linear target + adapter lifecycle |
| TRL | **SFT/DPO/GRPO 已验证** | 其他 trainer 逐项补 smoke |
| Accelerate / DeepSpeed | **当前 ZeRO smoke 矩阵已验证** | 不等于完整 PP/TP 生产训练 |
| verl | **待建立正式 adapter/recipe** | HF actor/rollout 是正确性起点 |
| OpenRLHF | **contract-ready，未验证** | 先走 HF actor；rollout backend 另行验证 |
| vLLM/SGLang rollout | **不是现成支持** | 它们会 hard-code 模型实现；需要 RWKV backend/plugin 或专用 rollout worker |

这也是“模型接口兼容”和“高性能 rollout engine 已适配”必须分开的原因：前者当前已有实证，后者不能靠 `trust_remote_code=True` 自动获得。

## 5. 可复现命令

```bash
MODEL=wangyue114514/rwkv7-g1d-0.1b-hf

python tests/test_native_trainer_smoke.py --model "$MODEL" --dtype fp32 --max-steps 2
python tests/test_native_sft_smoke.py --model "$MODEL" --dtype fp32 --max-steps 2
python tests/test_native_dpo_smoke.py --model "$MODEL" --dtype fp32 --max-steps 2
python tests/test_native_grpo_smoke.py --model "$MODEL" --dtype fp32 --max-steps 2
python tests/test_native_trainer_resume_smoke.py --model "$MODEL" --dtype fp32
python tests/test_native_peft_save_load_merge.py --model "$MODEL" --device cuda --dtype fp32 --steps 2
```

门禁不只是“进程退出 0”，还要包含：finite loss、非零梯度/参数变化、生成/奖励链路有效以及 save/load/resume 一致性。

## 6. 后续验收项

- [ ] 增加 `tests/test_native_dapo_smoke.py`，分别验证 asymmetric clipping、token-level loss、dynamic refill、overlong shaping 和非零参数更新。
- [ ] 增加 RLOO、KTO/ORPO/CPO 中至少一条代表性 smoke，验证“contract-ready”结论。
- [ ] 增加 `integrations/verl/` 的 RWKV-7 FSDP + HF-rollout 最小 recipe。
- [ ] 增加 verl actor/old/ref log-prob parity 门禁及 checkpoint resume。
- [ ] 为长 CoT 增加 selected-token/chunked log-prob 路径，避免长期物化完整 `[B,T,V]` logits。
- [ ] 在正确性通过后实现/接入 RWKV 专用高吞吐 rollout worker；在此之前不声称 vLLM/SGLang 原生支持。

## 结论

当前仓库已经用真实 Trainer/PEFT/TRL/DeepSpeed 调用证明：Native 不是只能被自有 benchmark 或 fused inference kernel 驱动，它具备标准 HF 训练与 RL 所需的模型表面。更准确的表述是：

> HF Trainer、PEFT、TRL SFT/DPO/GRPO 和当前 DeepSpeed ZeRO smoke 矩阵已经验证；大量基于 CausalLM logits/generation 的训练算法在模型 contract 上可接入，但每个新增算法和外部 rollout engine 仍需独立 smoke/长跑，不能把接口兼容直接写成生产级全覆盖。

**Comment by 123123213weqw** [comment]
当前 v0.9 clean reference 与 v0.10 optional Triton 的正式实测状态已经单独整理到 #144。

#144 包含：

- v0.9 reference 的 V100 正式 48-unit `lm_eval`；
- V100 100-step LoRA SFT/DPO/GRPO、adapter reload、resume 和 W&B offline 证据；
- forced Triton 在 V100/RTX 4080/RTX 4090 上各 48/48 `lm_eval`；
- 三卡 HF ecosystem、training fallback、SFT/DPO/GRPO smoke；
- 尚未通过的 FP16 full-model logits `max_abs <= 0.15` 门禁。

本 Issue 保留为 DAPO/verl/其他 RL 框架的接口与后续 roadmap；当前发布和验证结论以 #144 的固定 SHA/JSON evidence 为准。
